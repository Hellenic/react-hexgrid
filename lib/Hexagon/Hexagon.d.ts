import * as React from "react";
import { Hex } from "../models/Hex";
import { Point } from "../models/Point";
declare type H = {
    data?: any;
    state: {
        hex: Hex;
    };
    props: HexagonProps;
};
export declare type HexagonDragEventHandler<T = Element, AdditionalData = any> = (event: React.DragEvent<T>, h: H, additionalData?: AdditionalData) => void;
export declare type HexagonDragDropEventHandler<T = Element, AdditionalData = any> = (event: React.DragEvent<T>, h: H, additionalData: AdditionalData) => void;
export declare type HexagonMouseEventHandler<T = SVGGElement> = (event: React.MouseEvent<T, globalThis.MouseEvent>, h: H) => void;
export declare type HexagonProps = {
    q: number;
    r: number;
    s: number;
    rings?: number;
    fill?: string;
    fillUrl?: string;
    className?: string;
    cellStyle?: React.CSSProperties | undefined;
    data?: any;
    onMouseEnter?: HexagonMouseEventHandler;
    onMouseLeave?: HexagonMouseEventHandler;
    onClick?: HexagonMouseEventHandler;
    onDragStart?: HexagonDragEventHandler;
    onDragEnd?: HexagonDragEventHandler;
    onDragOver?: HexagonDragEventHandler;
    onDrop?: HexagonDragDropEventHandler<any, TargetProps>;
    onMouseOver?: HexagonMouseEventHandler;
    children?: React.ReactNode | React.ReactNode[];
};
declare type TargetProps = {
    hex: Hex;
    pixel: Point;
    data?: any;
    fill?: string;
    fillUrl?: string;
    className?: string;
};
/**
 * Renders a Hexagon cell at the given rqs-based coordinates.
 */
export declare function Hexagon(props: HexagonProps & Omit<React.SVGProps<SVGGElement>, "transform" | "onDragStart" | "onDragEnd" | "onDrop" | "onDragOver" | "onMouseEnter" | "onClick" | "onMouseOver" | "onMouseLeave">): JSX.Element;
export default Hexagon;
//# sourceMappingURL=Hexagon.d.ts.map