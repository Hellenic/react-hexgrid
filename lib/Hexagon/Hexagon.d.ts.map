{"version":3,"file":"Hexagon.d.ts","sourceRoot":"","sources":["../../src/Hexagon/Hexagon.tsx"],"names":[],"mappings":"AAAA,OAAO,KAAK,KAAK,MAAM,OAAO,CAAA;AAE9B,OAAO,EAAE,GAAG,EAAE,MAAM,eAAe,CAAA;AAGnC,OAAO,EAAE,KAAK,EAAE,MAAM,iBAAiB,CAAA;AAEvC,aAAK,CAAC,GAAG;IAAE,IAAI,CAAC,EAAE,GAAG,CAAC;IAAC,KAAK,EAAE;QAAE,GAAG,EAAE,GAAG,CAAA;KAAE,CAAC;IAAC,KAAK,EAAE,YAAY,CAAA;CAAE,CAAA;AAEjE,oBAAY,uBAAuB,CAAC,CAAC,GAAG,OAAO,EAAE,cAAc,GAAG,GAAG,IAAI,CACvE,KAAK,EAAE,KAAK,CAAC,SAAS,CAAC,CAAC,CAAC,EACzB,CAAC,EAAE,CAAC,EACJ,cAAc,CAAC,EAAE,cAAc,KAC5B,IAAI,CAAA;AAET,oBAAY,2BAA2B,CAAC,CAAC,GAAG,OAAO,EAAE,cAAc,GAAG,GAAG,IAAI,CAC3E,KAAK,EAAE,KAAK,CAAC,SAAS,CAAC,CAAC,CAAC,EACzB,CAAC,EAAE,CAAC,EACJ,cAAc,EAAE,cAAc,KAC3B,IAAI,CAAA;AAET,oBAAY,wBAAwB,CAAC,CAAC,GAAG,WAAW,IAAI,CACtD,KAAK,EAAE,KAAK,CAAC,UAAU,CAAC,CAAC,EAAE,UAAU,CAAC,UAAU,CAAC,EACjD,CAAC,EAAE,CAAC,KACD,IAAI,CAAA;AAET,oBAAY,YAAY,GAAG;IACzB,CAAC,EAAE,MAAM,CAAA;IACT,CAAC,EAAE,MAAM,CAAA;IACT,CAAC,EAAE,MAAM,CAAA;IACT,KAAK,CAAC,EAAE,MAAM,CAAA;IACd,IAAI,CAAC,EAAE,MAAM,CAAA;IACb,SAAS,CAAC,EAAE,MAAM,CAAA;IAClB,SAAS,CAAC,EAAE,KAAK,CAAC,aAAa,GAAG,SAAS,CAAA;IAC3C,IAAI,CAAC,EAAE,GAAG,CAAA;IACV,YAAY,CAAC,EAAE,wBAAwB,CAAA;IACvC,YAAY,CAAC,EAAE,wBAAwB,CAAA;IACvC,OAAO,CAAC,EAAE,wBAAwB,CAAA;IAClC,WAAW,CAAC,EAAE,uBAAuB,CAAA;IACrC,SAAS,CAAC,EAAE,uBAAuB,CAAA;IACnC,UAAU,CAAC,EAAE,uBAAuB,CAAA;IACpC,MAAM,CAAC,EAAE,2BAA2B,CAAC,GAAG,EAAE,WAAW,CAAC,CAAA;IACtD,WAAW,CAAC,EAAE,wBAAwB,CAAA;IACtC,QAAQ,CAAC,EAAE,KAAK,CAAC,SAAS,GAAG,KAAK,CAAC,SAAS,EAAE,CAAA;CAC/C,CAAA;AAED,aAAK,WAAW,GAAG;IACjB,GAAG,EAAE,GAAG,CAAA;IACR,KAAK,EAAE,KAAK,CAAA;IACZ,IAAI,CAAC,EAAE,GAAG,CAAA;IACV,IAAI,CAAC,EAAE,MAAM,CAAA;IACb,SAAS,CAAC,EAAE,MAAM,CAAA;CACnB,CAAA;AAED;;GAEG;AACH,wBAAgB,OAAO,CACrB,KAAK,EAAE,YAAY,GACjB,IAAI,CACF,KAAK,CAAC,QAAQ,CAAC,WAAW,CAAC,EACzB,WAAW,GACX,aAAa,GACb,WAAW,GACX,QAAQ,GACR,YAAY,GACZ,cAAc,GACd,SAAS,GACT,aAAa,GACb,cAAc,CACjB,eAuHJ;AAED,eAAe,OAAO,CAAA"}