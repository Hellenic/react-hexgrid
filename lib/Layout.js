var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx } from "react/jsx-runtime";
import * as React from "react";
import { Orientation } from "./models/Orientation";
import { Point } from "./models/Point";
const LAYOUT_FLAT = new Orientation(3.0 / 2.0, 0.0, Math.sqrt(3.0) / 2.0, Math.sqrt(3.0), 2.0 / 3.0, 0.0, -1.0 / 3.0, Math.sqrt(3.0) / 3.0, 0.0);
const LAYOUT_POINTY = new Orientation(Math.sqrt(3.0), Math.sqrt(3.0) / 2.0, 0.0, 3.0 / 2.0, Math.sqrt(3.0) / 3.0, -1.0 / 3.0, 0.0, 2.0 / 3.0, 0.5);
const defaultSize = new Point(10, 10);
const defaultOrigin = new Point(0, 0);
const defaultSpacing = 1.0;
const Context = React.createContext({
    layout: {
        size: defaultSize,
        orientation: LAYOUT_FLAT,
        origin: defaultOrigin,
        spacing: defaultSpacing,
    },
    points: "",
});
export function useLayoutContext() {
    const ctx = React.useContext(Context);
    return ctx;
}
/**
 * Calculates the points for a hexagon given the size, angle, and center
 * @param circumradius Radius of the Hexagon
 * @param angle Angle offset for the hexagon in radians
 * @param center Central point for the heaxagon
 * @returns Array of 6 points
 */
export function calculateCoordinates(circumradius, angle = 0, center = new Point(0, 0), rings = 0) {
    let c = [];
    if (rings <= 0) {
        for (let i = 0; i < 6; i++) {
            const x = circumradius * Math.cos((2 * Math.PI * i) / 6 + angle);
            const y = circumradius * Math.sin((2 * Math.PI * i) / 6 + angle);
            const point = new Point(center.x + x, center.y + y);
            c.push(point);
        }
    }
    else {
        c = [
            new Point(3.5, -6.0621779),
            new Point(7.0, -12.124356),
            new Point(14.0, -12.124356),
            new Point(17.5, -6.062177900000001),
            new Point(24.5, -6.062177800000001),
            new Point(28.0, 0.0),
            new Point(24.5, 6.0621775),
            new Point(28.0, 12.124355),
            new Point(24.5, 18.186533),
            new Point(17.5, 18.186533),
            new Point(14.0, 24.248711),
            new Point(7.0, 24.248711),
            new Point(3.5, 18.186533),
            new Point(-3.5, 18.186533),
            new Point(-7.0, 12.124355),
            new Point(-3.5, 6.0621775),
            new Point(-7.0, 0.0),
            new Point(-3.5, -6.0621778),
            new Point(3.5, -6.0621779)
        ];
    }
    return c;
}
/**
 * Provides LayoutContext for all descendands and renders child elements inside a <g> (Group) element
 */
export function Layout(_a) {
    var { size = defaultSize, flat = true, spacing = defaultSpacing, origin = defaultOrigin, children, className } = _a, rest = __rest(_a, ["size", "flat", "spacing", "origin", "children", "className"]);
    const orientation = flat ? LAYOUT_FLAT : LAYOUT_POINTY;
    const angle = flat ? 0 : Math.PI / 6;
    const cornerCoords = calculateCoordinates(size.x, angle);
    const points = cornerCoords.map((point) => `${point.x},${point.y}`).join(" ");
    const childLayout = Object.assign({}, rest, {
        orientation,
        size,
        origin,
        spacing,
    });
    return (_jsx(Context.Provider, Object.assign({ value: {
            layout: childLayout,
            points,
        } }, { children: _jsx("g", Object.assign({ className: className }, { children: children })) })));
}
export default Layout;
//# sourceMappingURL=Layout.js.map