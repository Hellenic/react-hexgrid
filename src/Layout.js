import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Orientation from './models/Orientation';
import Point from './models/Point';

export const LAYOUT_FLAT = new Orientation(3.0 / 2.0, 0.0, Math.sqrt(3.0) / 2.0, Math.sqrt(3.0),2.0 / 3.0, 0.0, -1.0 / 3.0, Math.sqrt(3.0) / 3.0, 0.0);
export const LAYOUT_POINTY = new Orientation(Math.sqrt(3.0), Math.sqrt(3.0) / 2.0, 0.0, 3.0 / 2.0, Math.sqrt(3.0) / 3.0, -1.0 / 3.0, 0.0, 2.0 / 3.0, 0.5);

class Layout extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    className: PropTypes.string,
    flat: PropTypes.bool,
    origin: PropTypes.object,
    size: PropTypes.object,
    spacing: PropTypes.number
  };

  static defaultProps = {
    size: new Point(10, 10),
    flat: true,
    spacing: 1.0,
    origin: new Point(0, 0)
  }

  getPointOffset(corner, orientation, size) {
    let angle = 2.0 * Math.PI * (corner + orientation.startAngle) / 6;
    return new Point(size.x * Math.cos(angle), size.y * Math.sin(angle));
  }

  // TODO Refactor
  calculateCoordinates(orientation) {
    const corners = [];
    const center = new Point(0, 0);
    const { size } = this.props;

    Array.from(new Array(6), (x, i) => {
      const offset = this.getPointOffset(i, orientation, size);
      const point = new Point(center.x + offset.x, center.y + offset.y);
      corners.push(point);
    });

    return corners;
  }

  render() {
    const { children, flat, className, ...rest } = this.props;
    const orientation = (flat) ? LAYOUT_FLAT : LAYOUT_POINTY;
    const cornerCoords = this.calculateCoordinates(orientation);
    const points = cornerCoords.map(point => `${point.x},${point.y}`).join(' ');
    const layout = {...rest, orientation};
    return (
      <LayoutProvider value={{ layout, points }}>
        <g className={className}>
          {children}
        </g>
      </LayoutProvider>
    );
  }
}

export const LayoutContext = React.createContext({
  layout: LAYOUT_FLAT,
  points: ""
})
export const { Provider: LayoutProvider, Consumer: LayoutConsumer } = LayoutContext;

export default Layout;
