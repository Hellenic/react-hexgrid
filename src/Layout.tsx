import * as React from "react"
import { Orientation } from "./models/Orientation"
import { Point } from "./models/Point"

export type Size = { x: number; y: number }

export type LayoutDimension = {
  size: Size
  orientation: Orientation
  origin: Size
  spacing: number
}
export type LayoutContextProps = {
  layout: LayoutDimension
  points: string
}

const LAYOUT_FLAT = new Orientation(
  3.0 / 2.0,
  0.0,
  Math.sqrt(3.0) / 2.0,
  Math.sqrt(3.0),
  2.0 / 3.0,
  0.0,
  -1.0 / 3.0,
  Math.sqrt(3.0) / 3.0,
  0.0,
)
const LAYOUT_POINTY = new Orientation(
  Math.sqrt(3.0),
  Math.sqrt(3.0) / 2.0,
  0.0,
  3.0 / 2.0,
  Math.sqrt(3.0) / 3.0,
  -1.0 / 3.0,
  0.0,
  2.0 / 3.0,
  0.5,
)
const defaultSize = new Point(10, 10)
const defaultOrigin = new Point(0, 0)
const defaultSpacing = 1.0

const Context = React.createContext<LayoutContextProps>({
  layout: {
    size: defaultSize,
    orientation: LAYOUT_FLAT,
    origin: defaultOrigin,
    spacing: defaultSpacing,
  },
  points: "",
})

export function useLayoutContext() {
  const ctx = React.useContext(Context)
  return ctx
}

/**
 * Calculates the points for a hexagon given the size, angle, and center
 * @param circumradius Radius of the Hexagon
 * @param angle Angle offset for the hexagon in radians
 * @param center Central point for the heaxagon
 * @returns Array of 6 points
 */

function calculateCoordinates(
  circumradius: number,
  angle: number = 0,
  center: Point = new Point(0, 0),
) {
  const corners: Point[] = []

  for (let i = 0; i < 6; i++) {
    const x = circumradius * Math.cos((2 * Math.PI * i) / 6 + angle)
    const y = circumradius * Math.sin((2 * Math.PI * i) / 6 + angle)
    const point = new Point(center.x + x, center.y + y)
    corners.push(point)
  }

  return corners
}

export type LayoutProps = {
  children:
    | React.ReactElement
    | React.ReactElement[]
    | JSX.Element
    | JSX.Element[]
  className?: string
  flat?: boolean
  origin?: any
  /* defines scale */
  size?: Size
  space?: number
  spacing?: number
}

/**
 * Provides LayoutContext for all descendands and renders child elements inside a <g> (Group) element
 */
export function Layout({
  size = defaultSize,
  flat = true,
  spacing = defaultSpacing,
  origin = defaultOrigin,
  children,
  className,
  ...rest
}: LayoutProps) {
  const orientation = flat ? LAYOUT_FLAT : LAYOUT_POINTY
  const angle = flat ? 0 : Math.PI / 6
  const cornerCoords = calculateCoordinates(size.x, angle)
  const points = cornerCoords.map((point) => `${point.x},${point.y}`).join(" ")
  const childLayout = Object.assign({}, rest, {
    orientation,
    size,
    origin,
    spacing,
  })

  return (
    <Context.Provider
      value={{
        layout: childLayout,
        points,
      }}
    >
      <g className={className}>{children}</g>
    </Context.Provider>
  )
}

export default Layout
