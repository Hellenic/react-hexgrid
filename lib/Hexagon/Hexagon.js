var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from "react";
import classNames from "classnames";
import { Hex } from "../models/Hex";
import { HexUtils } from "../HexUtils";
import { useLayoutContext, calculateCoordinates } from "../Layout";
import { Point } from "../models/Point";
/**
 * Renders a Hexagon cell at the given rqs-based coordinates.
 */
export function Hexagon(props) {
    // destructure props into their values
    const { q, r, s, rings, fill, cellStyle, className, children, onDragStart, onDragEnd, onDrop, onDragOver, onMouseEnter, onMouseLeave, onMouseOver, onClick, data, fillOpacity } = props, rest = __rest(props, ["q", "r", "s", "rings", "fill", "cellStyle", "className", "children", "onDragStart", "onDragEnd", "onDrop", "onDragOver", "onMouseEnter", "onMouseLeave", "onMouseOver", "onClick", "data", "fillOpacity"]);
    const { layout, points } = useLayoutContext();
    let q2 = q;
    let r2 = r;
    let s2 = s;
    if (rings == 1) {
        r2 -= 1;
    }
    else if (rings == 2) {
        q2 -= 2;
        r2 += 1;
    }
    const cornerCoords = calculateCoordinates(layout.size.x, 0, new Point(0, 0), rings);
    const ps = cornerCoords.map((point) => `${point.x},${point.y}`).join(" ");
    const { hex, pixel } = React.useMemo(() => {
        const hex = new Hex(q2, r2, s2);
        const pixel = HexUtils.hexToPixel(hex, layout);
        return {
            hex,
            pixel,
        };
    }, [q, r, s, layout]);
    // for backwards comapatbility
    const state = { hex };
    const fillId = fill ? `url(#${fill})` : undefined;
    const draggable = { draggable: true };
    return (_jsx("g", Object.assign({ className: classNames("hexagon-group", className), transform: `translate(${pixel.x}, ${pixel.y})` }, rest, draggable, { onDragStart: (e) => {
            if (onDragStart) {
                const targetProps = {
                    hex: hex,
                    pixel,
                    data: data,
                    fill: fill,
                    className: className,
                };
                e.dataTransfer.setData("hexagon", JSON.stringify(targetProps));
                onDragStart(e, { data, state, props });
            }
        }, onDragEnd: (e) => {
            if (onDragEnd) {
                e.preventDefault();
                const success = e.dataTransfer.dropEffect !== "none";
                onDragEnd(e, { state, props }, success);
            }
        }, onDrop: (e) => {
            if (onDrop) {
                e.preventDefault();
                const target = JSON.parse(e.dataTransfer.getData("hexagon"));
                onDrop(e, { data, state, props }, target);
            }
        }, onDragOver: (e) => {
            if (onDragOver) {
                onDragOver(e, { data, state, props });
            }
        }, onMouseEnter: (e) => {
            if (onMouseEnter) {
                onMouseEnter(e, { data, state, props });
            }
        }, onClick: (e) => {
            if (onClick) {
                onClick(e, { data, state, props });
            }
        }, onMouseOver: (e) => {
            if (onMouseOver) {
                onMouseOver(e, { data, state, props });
            }
        }, onMouseLeave: (e) => {
            if (onMouseLeave) {
                onMouseLeave(e, { data, state, props });
            }
        } }, { children: _jsxs("g", Object.assign({ className: "hexagon" }, { children: [_jsx("polygon", { points: ps, fill: fillId, style: cellStyle }), children] })) })));
}
export default Hexagon;
//# sourceMappingURL=Hexagon.js.map