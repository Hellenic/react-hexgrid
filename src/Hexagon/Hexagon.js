import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Hex from '../models/Hex';
import HexUtils from '../HexUtils';
import { LayoutConsumer } from '../Layout';

class Hexagon extends Component {
  static propTypes = {
    q: PropTypes.number.isRequired,
    r: PropTypes.number.isRequired,
    s: PropTypes.number.isRequired,
    layout: PropTypes.objectOf(PropTypes.any).isRequired,
    points: PropTypes.string.isRequired,
    fill: PropTypes.string,
    cellStyle: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.object
    ]),
    className: PropTypes.string,
    data: PropTypes.object,
    onMouseEnter: PropTypes.func,
    onMouseOver: PropTypes.func,
    onMouseLeave: PropTypes.func,
    onClick: PropTypes.func,
    onDragStart: PropTypes.func,
    onDragEnd: PropTypes.func,
    onDragOver: PropTypes.func,
    onDrop: PropTypes.func,
    children: PropTypes.node
  };

  static getDerivedStateFromProps(nextProps, prevState) {
    const { q, r, s, layout } = nextProps;
    const hex = new Hex(q, r, s);
    const pixel = HexUtils.hexToPixel(hex, layout);

    if (
      (!prevState.hex && !prevState.pixel)
      || !((HexUtils.equals(prevState.hex, hex))
        || (prevState.pixel.x === pixel.x && prevState.pixel.y === pixel.y))
    ) {
      return { hex, pixel };
    }
    return null;
  }

  state = { hex: {}, pixel: {} };

  onMouseEnter(e) {
    if (this.props.onMouseEnter) {
      this.props.onMouseEnter(e, this);
    }
  }
  onMouseOver(e) {
    if (this.props.onMouseOver) {
      this.props.onMouseOver(e, this);
    }
  }
  onMouseLeave(e) {
    if (this.props.onMouseLeave) {
      this.props.onMouseLeave(e, this);
    }
  }
  onClick(e) {
    if (this.props.onClick) {
      this.props.onClick(e, this);
    }
  }
  onDragStart(e) {
    if (this.props.onDragStart) {
      const targetProps = {
        ...this.state,
        data: this.props.data,
        fill: this.props.fill,
        className: this.props.className
      };
      e.dataTransfer.setData('hexagon', JSON.stringify(targetProps));
      this.props.onDragStart(e, this);
    }
  }
  onDragEnd(e) {
    if (this.props.onDragEnd) {
      e.preventDefault();
      const success = (e.dataTransfer.dropEffect !== 'none');
      this.props.onDragEnd(e, this, success);
    }
  }
  onDragOver(e) {
    if (this.props.onDragOver) {
      this.props.onDragOver(e, this);
    }
  }
  onDrop(e) {
    if (this.props.onDrop) {
      e.preventDefault();
      const target = JSON.parse(e.dataTransfer.getData('hexagon'));
      this.props.onDrop(e, this, target);
    }
  }
  render() {
    const { fill, cellStyle, className, points } = this.props;
    const { hex, pixel } = this.state;
    const fillId = (fill) ? `url(#${fill})` : null;
    return (
      <g
        className={classNames('hexagon-group', className)}
        transform={`translate(${pixel.x}, ${pixel.y})`}
        draggable="true"
        onMouseEnter={e => this.onMouseEnter(e)}
        onMouseOver={e => this.onMouseOver(e)}
        onMouseLeave={e => this.onMouseLeave(e)}
        onClick={e => this.onClick(e)}
        onDragStart={e => this.onDragStart(e)}
        onDragEnd={e => this.onDragEnd(e)}
        onDragOver={e => this.onDragOver(e)}
        onDrop={e => this.onDrop(e)}
      >
        <g className="hexagon">
          <polygon points={points} fill={fillId} style={cellStyle} />
          {this.props.children}
        </g>
      </g>
    );
  }
}

export default props => <LayoutConsumer>{({layout, points}) => <Hexagon layout={layout} points={points} {...props} />}</LayoutConsumer>;
