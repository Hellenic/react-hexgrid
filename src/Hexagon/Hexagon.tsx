import * as React from "react"
import classNames from "classnames"
import { Hex } from "../models/Hex"
import { HexUtils } from "../HexUtils"
import { useLayoutContext, calculateCoordinates } from "../Layout"
import { Point } from "../models/Point"

type H = { data?: any; state: { hex: Hex }; props: HexagonProps }

export type HexagonDragEventHandler<T = Element, AdditionalData = any> = (
  event: React.DragEvent<T>,
  h: H,
  additionalData?: AdditionalData,
) => void

export type HexagonDragDropEventHandler<T = Element, AdditionalData = any> = (
  event: React.DragEvent<T>,
  h: H,
  additionalData: AdditionalData,
) => void

export type HexagonMouseEventHandler<T = SVGGElement> = (
  event: React.MouseEvent<T, globalThis.MouseEvent>,
  h: H,
) => void

export type HexagonProps = {
  q: number
  r: number
  s: number
  rings?: number
  fill?: string
  className?: string
  cellStyle?: React.CSSProperties | undefined
  data?: any
  onMouseEnter?: HexagonMouseEventHandler
  onMouseLeave?: HexagonMouseEventHandler
  onClick?: HexagonMouseEventHandler
  onDragStart?: HexagonDragEventHandler
  onDragEnd?: HexagonDragEventHandler
  onDragOver?: HexagonDragEventHandler
  onDrop?: HexagonDragDropEventHandler<any, TargetProps>
  onMouseOver?: HexagonMouseEventHandler
  children?: React.ReactNode | React.ReactNode[]
}

type TargetProps = {
  hex: Hex
  pixel: Point
  data?: any
  fill?: string
  className?: string
}

/**
 * Renders a Hexagon cell at the given rqs-based coordinates.
 */
export function Hexagon(
  props: HexagonProps &
    Omit<
      React.SVGProps<SVGGElement>,
      | "transform"
      | "onDragStart"
      | "onDragEnd"
      | "onDrop"
      | "onDragOver"
      | "onMouseEnter"
      | "onClick"
      | "onMouseOver"
      | "onMouseLeave"
    >,
) {
  // destructure props into their values
  const {
    q,
    r,
    s,
    rings,
    fill,
    cellStyle,
    className,
    children,
    onDragStart,
    onDragEnd,
    onDrop,
    onDragOver,
    onMouseEnter,
    onMouseLeave,
    onMouseOver,
    onClick,
    data,
    fillOpacity,
    ...rest
  } = props

  const { layout, points } = useLayoutContext()
layout.orientation
  const cornerCoords = calculateCoordinates(layout.size.x, 0, new Point(0, 0), rings)
  const ps = cornerCoords.map((point) => `${point.x},${point.y}`).join(" ")

  const { hex, pixel } = React.useMemo(() => {
    const hex = new Hex(q, r, s)
    const pixel = HexUtils.hexToPixel(hex, layout)
    return {
      hex,
      pixel,
    }
  }, [q, r, s, layout])

  // for backwards comapatbility
  const state = { hex }

  const fillId = fill ? `url(#${fill})` : undefined
  const draggable = { draggable: true } as any
  return (
    <g
      className={classNames("hexagon-group", className)}
      transform={`translate(${pixel.x}, ${pixel.y})`}
      {...rest}
      {...draggable}
      onDragStart={(e) => {
        if (onDragStart) {
          const targetProps: TargetProps = {
            hex: hex,
            pixel,
            data: data,
            fill: fill,
            className: className,
          }
          e.dataTransfer.setData("hexagon", JSON.stringify(targetProps))
          onDragStart(e, { data, state, props })
        }
      }}
      onDragEnd={(e) => {
        if (onDragEnd) {
          e.preventDefault()
          const success = e.dataTransfer.dropEffect !== "none"
          onDragEnd(e, { state, props }, success)
        }
      }}
      onDrop={(e) => {
        if (onDrop) {
          e.preventDefault()
          const target = JSON.parse(e.dataTransfer.getData("hexagon"))
          onDrop(e, { data, state, props }, target)
        }
      }}
      onDragOver={(e) => {
        if (onDragOver) {
          onDragOver(e, { data, state, props })
        }
      }}
      onMouseEnter={(e) => {
        if (onMouseEnter) {
          onMouseEnter(e, { data, state, props })
        }
      }}
      onClick={(e) => {
        if (onClick) {
          onClick(e, { data, state, props })
        }
      }}
      onMouseOver={(e) => {
        if (onMouseOver) {
          onMouseOver(e, { data, state, props })
        }
      }}
      onMouseLeave={(e) => {
        if (onMouseLeave) {
          onMouseLeave(e, { data, state, props })
        }
      }}
    >
      <g className="hexagon">
        <polygon points={ps} fill={fillId} style={cellStyle} />
        {children}
      </g>
    </g>
  )
}

export default Hexagon
