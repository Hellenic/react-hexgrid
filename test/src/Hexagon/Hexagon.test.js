import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import Layout from '../../../src/Layout';
import Hexagon from '../../../src/Hexagon/Hexagon';

test('Hexagon should render correctly with default props', () => {
  const tree = renderer.create(
    <Layout
      className={'test1'}
      size={{ x: 6, y: 6 }}
      flat={false}
      spacing={1.1}
      origin={{ x: 0, y: 0 }}
    >
      <Hexagon
        q={0}
        r={0}
        s={0}
      >
        <div>child</div>
      </Hexagon>
    </Layout>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Hexagon should work', () => {
  let playDoh;
  const wrapper = mount(
    <Layout
      className={'layout'}
      size={{ x: 6, y: 6 }}
      flat={false}
      spacing={1.1}
      origin={{ x: 0, y: 0 }}
    >
      <Hexagon
        q={0}
        r={0}
        s={0}
        fill={'#333'}
        className={'test1'}
        data={{ a: 'b' }}
        onMouseEnter={() => { playDoh = 'mouseEnter'; }}
        onMouseOver={() => { playDoh = 'mouseOver'; }}
        onMouseLeave={() => { playDoh = 'mouseLeave'; }}
        onClick={() => { playDoh = 'click'; }}
        onDragStart={() => { playDoh = 'dragStart'; }}
        onDragEnd={() => { playDoh = 'dragEnd'; }}
        onDragOver={() => { playDoh = 'dragOver'; }}
        onDrop={() => { playDoh = 'drop'; }}
      >
        <div>child</div>
      </Hexagon>
    </Layout>,
  );
  expect(wrapper.find('g.test1').length).toBe(1);

  wrapper.find('g.test1').simulate('mouseEnter');
  expect(playDoh).toBe('mouseEnter');

  wrapper.find('g.test1').simulate('mouseOver');
  expect(playDoh).toBe('mouseOver');

  wrapper.find('g.test1').simulate('mouseLeave');
  expect(playDoh).toBe('mouseLeave');

  wrapper.find('g.test1').simulate('click');
  expect(playDoh).toBe('click');

  wrapper.find('g.test1').simulate('dragStart', { dataTransfer: { setData: () => {} } });
  expect(playDoh).toBe('dragStart');

  wrapper.find('g.test1').simulate('dragEnd', { dataTransfer: { setData: () => {} } });
  expect(playDoh).toBe('dragEnd');

  wrapper.find('g.test1').simulate('dragOver');
  expect(playDoh).toBe('dragOver');

  wrapper.find('g.test1').simulate('drop', { dataTransfer: { getData: data => JSON.stringify({ data }) } });
  expect(playDoh).toBe('drop');
});

test('Hexagon should work', () => {
  let playDoh;
  const wrapper2 = mount(
    <Layout
      className={'layout'}
      size={{ x: 6, y: 6 }}
      flat={false}
      spacing={1.1}
      origin={{ x: 0, y: 0 }}
    >
      <Hexagon
        q={0}
        r={0}
        s={0}
        className={'test2'}
      >
        <div>child</div>
      </Hexagon>
    </Layout>,
  );
  console.log(wrapper2.find('g.test2').length);
  expect(wrapper2.find('g.test2').length).toBe(1);

  wrapper2.find('g.test2').simulate('mouseEnter');
  expect(playDoh).toBe(undefined);

  wrapper2.find('g.test2').simulate('mouseOver');
  expect(playDoh).toBe(undefined);

  wrapper2.find('g.test2').simulate('mouseLeave');
  expect(playDoh).toBe(undefined);

  wrapper2.find('g.test2').simulate('click');
  expect(playDoh).toBe(undefined);

  wrapper2.find('g.test2').simulate('dragStart');
  expect(playDoh).toBe(undefined);

  wrapper2.find('g.test2').simulate('dragEnd');
  expect(playDoh).toBe(undefined);

  wrapper2.find('g.test2').simulate('dragOver');
  expect(playDoh).toBe(undefined);

  wrapper2.find('g.test2').simulate('drop');
  expect(playDoh).toBe(undefined);
});
