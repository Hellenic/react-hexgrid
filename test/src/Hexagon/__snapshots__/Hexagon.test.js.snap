// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hexagon should render correctly with default props 1`] = `
<g
  className="test1"
>
  <g
    className="hexagon-group"
    draggable="true"
    onClick={[Function]}
    onDragEnd={[Function]}
    onDragOver={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    transform="translate(0, 0)"
  >
    <g
      className="hexagon"
    >
      <polygon
        fill={null}
        points="5.196152422706632,2.9999999999999996 3.6739403974420594e-16,6 -5.196152422706632,2.9999999999999996 -5.196152422706633,-2.9999999999999982 -1.102182119232618e-15,-6 5.19615242270663,-3.0000000000000027"
      />
      <div>
        child
      </div>
    </g>
  </g>
</g>
`;
